abci/src/application/kvstore.rs

```diff
@@ -0,0 +1,274 @@
+//! In-memory key/value store ABCI application.
+
+use crate::codec::{encode_varint, MAX_VARINT_LENGTH};
+use crate::{Application, Error, Result};
+use bytes::BytesMut;
+use std::collections::HashMap;
+use std::sync::mpsc::{channel, Receiver, Sender};
+use tendermint_proto::abci::{
+    Event, EventAttribute, RequestCheckTx, RequestDeliverTx, RequestInfo, RequestQuery,
+    ResponseCheckTx, ResponseCommit, ResponseDeliverTx, ResponseInfo, ResponseQuery,
+};
+use tracing::{debug, info};
+
+/// In-memory, hashmap-backed key/value store ABCI application.
+///
+/// This structure effectively just serves as a handle to the actual key/value
+/// store - the [`KeyValueStoreDriver`].
+#[derive(Debug, Clone)]
+pub struct KeyValueStoreApp {
+    cmd_tx: Sender<Command>,
+}
+
+impl KeyValueStoreApp {
+    /// Constructor.
+    pub fn new() -> (Self, KeyValueStoreDriver) {
+        let (cmd_tx, cmd_rx) = channel();
+        (Self { cmd_tx }, KeyValueStoreDriver::new(cmd_rx))
+    }
+
+    /// Attempt to retrieve the value associated with the given key.
+    pub fn get<K: AsRef<str>>(&self, key: K) -> Result<(i64, Option<String>)> {
+        let (result_tx, result_rx) = channel();
+        channel_send(
```

Comment:
Could you explain why you've chosen this concurrency model (communication via channels) over something like a mutex & hashmap? _As someone who's learning Rust it will be useful to me._

Reply:
This may be a personal preference/readability thing :grin: I'd say there are pros and cons to each approach. I've always tended to prefer an actor-style model for managing state in concurrent systems (explicit-lock-free state encapsulation with message passing and pattern matching on messages), which I've tried to implement here in a lightweight way. It's easier for me to reason about such state machines in general ([like this](https://github.com/informalsystems/tendermint-rs/pull/794/files#diff-d3dc4894c066ca996f470f741a50cd983e04000bd8dce1960ce0adccce6dad72R211)).

When it comes to contention, I'd imagine using mutexes would push bottlenecks down the stack towards the networking layer. With the actor-based model it'd be able to leverage memory (in the unbounded channel) to avoid pushing the bottleneck towards the networking layer. On more constrained devices, this channel-based approach wouldn't provide any benefit for the additional complexity cost, so mutexes would probably be the way to go.

Practically both models would probably work the same for this specific application, although I'm not sure which one would provide better performance (I'd have to benchmark to find out). I'd imagine that the mutex-based approach would probably provide better performance, speed-wise, as long as the app can keep up with incoming requests.